### ISSUE TITLE
Interaction ids get conflated after iterated give

### ISSUE THREAD

**Issue by sattlerc** [agda/agda#4252]
Original title was: Internal error after iterated give: module application with postfix operator

On master (36b9448).

```agda
postulate
  admit : ∀ {i} {X : Set i} → X
  Id : ∀ {A : Set} (a : A) → Set

record Unit : Set where
  constructor unit

record Σ (A : Set) (B : A → Set) : Set where
  field
    fst : A
    snd : B fst
open Σ

Equiv : (A B : Set) → Set
Equiv A B = Σ (A → B) (λ _ → B → A)

equiv : {A B : Set} (f : A → B) → Equiv A B
equiv f .fst = f
equiv f .snd = admit

_inv : {A B : Set} → Equiv A B → Equiv B A
_inv e .fst = e .snd
_inv e .snd = admit

postulate
  A X : Set
  a : A
  left : {A B : Set} (P : B → Set) (e : Equiv A B) → Equiv (Σ A (λ x → P (e .fst x))) (Σ B P)
  _noop : {A : Set} → A → A
  Σ-Unit-l : {B : Unit → Set} → Equiv (Σ Unit B) (B unit)
  R : (U V : Set) → Set₁
  foo : {A B C : Set} → Equiv B C → Equiv B A → R A C

module M (U V : Set) (p : R U V) where

module R = M
  (Σ (Σ A (λ _ → A)) (λ x → Id (x .fst)))
  (Id a)
  (foo Σ-Unit-l ({!left _ ?!} noop)) -- Give, then write '? inv' in hole, then give.
```
Giving the expression in the hole, then writing '? inv' in the new hole, and giving again yields the following internal error:
```
An internal error has occurred. Please report this as a bug.
Location of the error: src/full/Agda/TypeChecking/Records.hs:609
```
Note that the internal error does not appear with any of the following modifications:
* giving '? inv' via giving in the empty hole and then writing in the 'expression to give' prompt,
* replacing the expression with 'left _ {!? inv!}' and giving twice,
* turning either of the postfix operators into functions.

**Comment by sattlerc** [comment]
I reduced the test case, arriving at the following.
```agda
postulate
  A Z : Set

record R : Set where
  field
    a : A

record R' : Set where
  field
    a : A

_post : R' → R
_post e .R.a = e .R'.a

postulate
  Y : A → Set
  hum : (r : R) → Y (r .R.a)
  a : A
  pre_ : Y a → Z

foo : Z
foo = pre {!hum ?!} -- Give, then write '? post', then give.
```
The output is:
```
src/full/Agda/TypeChecking/Reduce/Fast.hs:896:19-53: Non-exhaustive patterns in Just n
```
With --no-fast-reduce, the output is:
```
An internal error has occurred. Please report this as a bug.
Location of the error: src/full/Agda/TypeChecking/Substitute.hs:133
```

It is possible to reduce the test case even further while retaining incorrect behaviour.
```Agda
postulate
  A Y Z : Set

record R : Set where
  constructor _post
  field
    a : A

postulate
  hum : R → Y
  pre_ : Y → Z

foo : Z
foo = pre {!hum ?!} -- Give, then write '? post', then give.
```
Output:
```
Failed to give (? post).
```
However, upon reloading and giving again, `? post` is accepted.

Again, turning either of the operators into functions causes the error to disappear. So it must have something to do with mixfix operators.

**Comment by sattlerc** [comment]
Note also that it is possible to give `(? post)`, even when `? post` is rejected.

**Comment by sattlerc** [comment]
Here is a further reduction.
```agda
postulate
  A B : Set
  _p : A → B

foo : B
foo = {!?!} -- Give, then write '? p', then give.
```
Output:
```
Failed to give ? p
```
Any takers?

**Comment by sattlerc** [comment]
Running a diff on `-v100` between a good case (first, giving `(? p)`) and a bad case (second, giving `? p`), the first essential difference is as follows:
```
giving expression ? p
App (AppInfo {appRange = Bug.agda:7,10-13, appOrigin = UserWritten, appParens = PreferParenless}) (Def Bug._p) (Arg {argInfo = ArgInfo {argInfoHiding = NotHidden, argInfoModality = Modality {modRelevance = Relevant, modQuantity = Quantityω QωInferred, modCohesion = Continuous}, argInfoOrigin = UserWritten, argInfoFreeVariables = UnknownFVs}, unArg = Named {nameOf = Nothing, namedThing = QuestionMark (MetaInfo {metaRange = Bug.agda:7,10-11, metaScope = ScopeInfo {_scopeCurrent = , _scopeModules = fromList [(,Scope {scopeName = , scopeParents = [], scopeNameSpaces = [(PrivateNS,NameSpace {nsNames = fromList [], nsModules = fromList [], nsInScope = fromList []}),(PublicNS,NameSpace {nsNames = fromList [], nsModules = fromList [], nsInScope = fromList []}),(ImportedNS,NameSpace {nsNames = fromList [], nsModules = fromList [], nsInScope = fromList []})], scopeImports = fromList [], scopeDatatypeModule = Nothing})], _scopeVarsToBind = [], _scopeLocals = [], _scopePrecedence = [], _scopeInverseName = fromList [], _scopeInverseModule = fromList [], _scopeInScope = fromList [], _scopeFixities = fromList [], _scopePolarities = fromList []}, metaNumber = Nothing, metaNameSuggestion = ""}) (InteractionId {interactionId = 2})}})
```
versus
```
giving expression ? p
App (AppInfo {appRange = Bug.agda:7,9-12, appOrigin = UserWritten, appParens = PreferParenless}) (Def Bug._p) (Arg {argInfo = ArgInfo {argInfoHiding = NotHidden, argInfoModality = Modality {modRelevance = Relevant, modQuantity = Quantityω QωInferred, modCohesion = Continuous}, argInfoOrigin = UserWritten, argInfoFreeVariables = UnknownFVs}, unArg = Named {nameOf = Nothing, namedThing = QuestionMark (MetaInfo {metaRange = Bug.agda:7,9-10, metaScope = ScopeInfo {_scopeCurrent = , _scopeModules = fromList [(,Scope {scopeName = , scopeParents = [], scopeNameSpaces = [(PrivateNS,NameSpace {nsNames = fromList [], nsModules = fromList [], nsInScope = fromList []}),(PublicNS,NameSpace {nsNames = fromList [], nsModules = fromList [], nsInScope = fromList []}),(ImportedNS,NameSpace {nsNames = fromList [], nsModules = fromList [], nsInScope = fromList []})], scopeImports = fromList [], scopeDatatypeModule = Nothing})], _scopeVarsToBind = [], _scopeLocals = [], _scopePrecedence = [], _scopeInverseName = fromList [], _scopeInverseModule = fromList [], _scopeInScope = fromList [], _scopeFixities = fromList [], _scopePolarities = fromList []}, metaNumber = Nothing, metaNameSuggestion = ""}) (InteractionId {interactionId = 1})}})
```
The essential difference is at the line end: `interactionId = 2` vs. `interactionId = 1`. So it looks like interaction ids get conflated.

Indeed, if in the previous test case you replace `{!?!}` by `{! ?!}` or insert an additional space before the new hole after giving for the first time, there is no failure. There is also no failure if you replace `? p` by `{- only to change range of ? -} ? p` when giving the second time. Note that it is not enough to give ` ? p` (with a space before '?' not preserved by GitHub's markup): the prefix space seems to get trimmed for determining the range.

That makes me suspect the conflation of interaction ids has something to in `registerInteractionPoint` (TypeChecking.Monad.MetaVars) with unjustified calls to the following function, which reuses interaction ids based on range:
https://github.com/agda/agda/blob/a1d5ce5963e8bfa712c219ba8935977bb5de1355/src/full/Agda/TypeChecking/Monad/MetaVars.hs#L380-L387

**Comment by sattlerc** [comment]
Here is an even simpler test case.
```agda
A : Set
A = {!?!} -- Give, then write '?', then give again.
```
Output:
```
Failed to give ?
```

**Comment by sattlerc** [comment]
Related: #3000

**Comment by sattlerc** [comment]
See #4495 for another test case.

**Comment by UlfNorell** [comment]
Fix to #3000 also fixed this.
